libass/ass_directwrite.c

```diff
@@ -625,17 +629,46 @@ static void add_font_face(IDWriteFontFace *face, ASS_FontProvider *provider
 {
     ASS_FontProviderMetaData meta = {0};
 
-    FontPrivate *font_priv = (FontPrivate *) calloc(1, sizeof(*font_priv));
-    if (!font_priv) {
-        IDWriteFontFace_Release(face);
-        return;
+    IDWriteFontFace3 *face3;
+    HRESULT hr = IDWriteFontFace_QueryInterface(face, &IID_IDWriteFontFace3,
+                                                (void **) &face3);
+    if (SUCCEEDED(hr) && face3) {
+        bool success = get_font_info_IDWriteFontFace3(face3, &meta);
+        IDWriteFontFace3_Release(face3);
+        if (!success)
+            goto cleanup;
     }
+
+    FontPrivate *font_priv = (FontPrivate *) calloc(1, sizeof(*font_priv));
+    if (!font_priv)
+        goto cleanup;
+
     font_priv->face = face;
+    face = NULL;
+
 #if DESKTOP_API
     font_priv->shared_hdc = hdc_retain(shared_hdc);
 #endif
 
     ass_font_provider_add_font(provider, &meta, NULL, 0, font_priv);
+
+cleanup:
+    if (meta.families) {
```

Comment:
I've seen such cleanup code in several places. Maybe it should be factored out too?

Reply:
It’s slightly different in each place, which makes it annoying to unify. Also, this very PR actually removes it from `ass_coretext`.

I think I’ll try to make a separate PR that will remove redundant memory allocations in `ass_font_provider_add_font` and hopefully get rid of this duplicated cleanup code altogether.